File: Data Structures & Algorithms/delete-node-in-a-bst/submission-11.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
func deleteNode(root *TreeNode, key int) *TreeNode {
    if root == nil {
        return nil
    }

    var parent *TreeNode
    curr := root

    for curr != nil && curr.Val != key {
        parent = curr

        if key < curr.Val {
            curr = curr.Left
        } else {
            curr = curr.Right
        }
    }

    if curr == nil {
        return root
    }

    var replacement *TreeNode
    if curr.Left == nil {
        replacement = curr.Right
    } else if curr.Right == nil {
        replacement = curr.Left
    } else {
        successorParent := curr
        successor := curr.Right

        for successor.Left != nil {
            successorParent = successor
            successor = successor.Left
        }

        curr.Val = successor.Val

        if successorParent == curr {
            successorParent.Right = successor.Right
        } else {
            successorParent.Left = successor.Right
        }

        return root
    }

    if parent == nil {
        return replacement
    } else if parent.Left == curr {
        parent.Left = replacement
    } else {
        parent.Right = replacement
    }

    return root
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-11.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        q = deque([root])
        result = []

        while q:
            q_len = len(q)
            last = q_len-1
            
            for i in range(q_len):
                node = q.popleft()

                # we need the last element at each level
                if i == last:
                    result.append(node.val)
                
                if node.left:
                    q.append(node.left)
                
                if node.right:
                    q.append(node.right)
        
        return result
